stream.go

```diff
@@ -0,0 +1,331 @@
+	"context"
+	"errors"
+	"io"
+	"sync"
+	"time"
+
+	"go.uber.org/atomic"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/anypb"
+
+	"github.com/livekit/psrpc/internal"
+)
+
+type Stream[SendType, RecvType proto.Message] interface {
+	Channel() <-chan *Response[RecvType]
+	Send(SendType) error
+	Close(error) error
+	Err() error
+}
+
+type ServerStream[SendType, RecvType proto.Message] interface {
+	Stream[SendType, RecvType]
+	Hijack()
+}
+
+type ClientStream[SendType, RecvType proto.Message] interface {
+	Stream[SendType, RecvType]
+}
+
+type streamAdapter interface {
+	send(ctx context.Context, msg *internal.Stream) error
+	close(streamID string)
+}
+
+type clientStream struct {
+	c     *RPCClient
+	rpc   string
+	topic string
+}
+
+func (s *clientStream) send(ctx context.Context, msg *internal.Stream) (err error) {
+	if err = s.c.bus.Publish(ctx, getStreamServerChannel(s.c.serviceName, s.rpc, s.topic), msg); err != nil {
+		err = NewError(Internal, err)
+	}
+	return
+}
+
+func (s *clientStream) close(streamID string) {
+	s.c.mu.Lock()
+	delete(s.c.streamChannels, streamID)
+	s.c.mu.Unlock()
+}
+
+type serverStream[SendType, RecvType proto.Message] struct {
+	h      *streamRPCHandlerImpl[SendType, RecvType]
+	s      *RPCServer
+	nodeID string
+}
+
+func (s *serverStream[RequestType, ResponseType]) send(ctx context.Context, msg *internal.Stream) (err error) {
+	if err = s.s.bus.Publish(ctx, getStreamChannel(s.s.serviceName, s.nodeID), msg); err != nil {
+		err = NewError(Internal, err)
+	}
+	return
+}
+
+func (s *serverStream[RequestType, ResponseType]) close(streamID string) {
+	s.h.mu.Lock()
+	delete(s.h.streams, streamID)
+	s.h.mu.Unlock()
+	s.h.handling.Dec()
+}
+
+type streamHandler[SendType, RecvType proto.Message] struct {
+	*streamImpl[SendType, RecvType]
+}
+
+func (h *streamHandler[SendType, RecvType]) Recv(msg proto.Message, err error) {
+	h.recv(msg, err)
+}
+
+func (h *streamHandler[SendType, RecvType]) Send(msg proto.Message, opts ...StreamOption) error {
+	return h.send(msg)
+}
+
+func (h *streamHandler[SendType, RecvType]) Close(cause error) error {
+	return h.close(cause)
+}
+
+type streamImpl[SendType, RecvType proto.Message] struct {
+	streamOpts
+	adapter   streamAdapter
+	handler   StreamHandler
+	ctx       context.Context
+	cancelCtx context.CancelFunc
+	recvChan  chan *Response[RecvType]
+	streamID  string
+	mu        sync.Mutex
+	recvQueue list.List
+	hijacked  bool
+	pending   atomic.Int32
+	acks      map[string]chan struct{}
+	err       error
+	closeOnce sync.Once
+}
+
+func (s *streamImpl[SendType, RecvType]) handleStream(is *internal.Stream) error {
+	switch b := is.Body.(type) {
+	case *internal.Stream_Ack:
+		s.mu.Lock()
+		ack, ok := s.acks[is.RequestId]
+		delete(s.acks, is.RequestId)
+		s.mu.Unlock()
+
+		if ok {
+			close(ack)
+		}
+
+	case *internal.Stream_Message:
+		v, err := b.Message.Message.UnmarshalNew()
+		if err != nil {
+			err = NewError(MalformedRequest, err)
```

Comment:
maybe this should be returned to the sender?

Reply:
I agree any error in sending should be returned to sender. Might make sense for Recv to be (value, ok) to match how channels work